File: examples/go/organization_example.go
package main

import (
	"github.com/castai/pulumi-castai/sdk/go/castai/organization"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

// runOrganizationExample demonstrates creating organization members and groups.
// Note: This function is not 'main' to avoid conflicts if other .go files
// in this directory also have a 'main' function. You might need to call
// this from your actual main entry point or run this file standalone.
func runOrganizationExample(ctx *pulumi.Context) error {
	// Note: Provider configuration (API token) is typically handled via environment
	// variables (e.g., CASTAI_API_TOKEN) or Pulumi config.

	// Manage organization members directly
	// This resource manages the entire list of members for the specified roles.
	// Use with caution, as it will overwrite existing members not specified here.
	orgMembers, err := organization.NewMembers(ctx, "example-org-members", &organization.MembersArgs{
		Members: organization.MembersMemberArray{
			&organization.MembersMemberArgs{
				Email: pulumi.String("admin.user@example.com"),
				Role:  pulumi.String("admin"),
			},
			&organization.MembersMemberArgs{
				Email: pulumi.String("dev.user@example.com"),
				Role:  pulumi.String("editor"),
			},
			&organization.MembersMemberArgs{
				Email: pulumi.String("view.user@example.com"),
				Role:  pulumi.String("viewer"),
			},
		},
	})
	if err != nil {
		return err
	}

	// Create an organization group
	orgGroup, err := organization.NewGroup(ctx, "example-devops-group", &organization.GroupArgs{
		Name:        pulumi.String("DevOps Team"),
		Description: pulumi.String("Group for DevOps engineers managing infrastructure"),
		MemberEmails: pulumi.StringArray{
			pulumi.String("devops1@example.com"),
			pulumi.String("devops2@example.com"),
			// You can add users who might also be managed directly above,
			// group membership is additive.
			pulumi.String("dev.user@example.com"),
		},
	})
	if err != nil {
		return err
	}

	ctx.Export("org_group_id", orgGroup.ID())
	ctx.Export("org_group_name", orgGroup.Name)

	// Reference the members resource to ensure it's created, even if not exported
	ctx.Export("org_members_id_ref", orgMembers.ID())

	// Note: For managing members, castai.organization.Members provides a declarative
	// way to set the *entire* list of members per role.
	// For adding/removing individual members without affecting others,
	// other mechanisms or resources might be needed depending on the provider's capabilities.
	// Similarly, castai.organization.Group manages the member list declaratively.

	return nil
}

// Example of how you might call this from your main function:
/*
func main() {
	pulumi.Run(func(ctx *pulumi.Context) error {
		if err := runOrganizationExample(ctx); err != nil {
			return err
		}
		// ... potentially call other example functions ...
		return nil
	})
}
*/

File: examples/python/organization_example.py
"""
CAST AI Organization Example for Python

This example demonstrates how to manage organization members and groups with CAST AI.
"""

import pulumi
import pulumi_castai as castai
import os
from typing import Dict, List, Optional, Any
from pydantic import BaseModel, Field

# Import the CAST AI provider and models
from models import Provider as ProviderModel

# Initialize the CAST AI provider
api_token = os.environ.get("CASTAI_API_TOKEN", "your-api-token-here")
provider_config = ProviderModel(api_token=api_token)
provider = castai.Provider("castai-provider",
    api_token=provider_config.api_token,
    api_url=provider_config.api_url
)

# Pydantic models for Organization resources
class OrganizationMember(BaseModel):
    """Organization member configuration."""
    email: str
    role: str

class OrganizationMembers(BaseModel):
    """Full organization members configuration."""
    members: List[OrganizationMember]

class OrganizationGroup(BaseModel):
    """Organization group configuration."""
    name: str
    description: Optional[str] = None
    member_emails: List[str]

# Create organization members configuration
org_members_model = OrganizationMembers(
    members=[
        OrganizationMember(
            email="admin.user@example.com",
            role="admin",
        ),
        OrganizationMember(
            email="dev.user@example.com",
            role="editor",
        ),
         OrganizationMember(
            email="view.user@example.com",
            role="viewer",
        ),
    ]
)

# Create organization group configuration
org_group_model = OrganizationGroup(
    name="DevOps Team",
    description="Group for DevOps engineers managing infrastructure",
    member_emails=[
        "devops1@example.com",
        "devops2@example.com",
        # You can add users who might also be managed directly above,
        # group membership is additive.
        "dev.user@example.com",
    ]
)

# Convert models to dictionaries
org_members_args = org_members_model.dict(exclude_none=True)
org_group_args = org_group_model.dict(exclude_none=True)

# Create resources with proper error handling
try:
    # Using direct Members resource
    org_members = castai.Members(
        "example-org-members",
        **org_members_args,
        opts=pulumi.ResourceOptions(provider=provider)
    )
    
    # Using direct Group resource
    org_group = castai.Group(
        "example-devops-group",
        **org_group_args,
        opts=pulumi.ResourceOptions(provider=provider)
    )
    
    # Export the group information
    pulumi.export("org_group_id", org_group.id)
    pulumi.export("org_group_name", org_group_model.name)
    
except Exception as e:
    pulumi.log.warn(f"Failed to create organization resources: {str(e)}")
    # Export the models as demonstration
    pulumi.export("org_members_model", org_members_args)
    pulumi.export("org_group_model", org_group_args)
    pulumi.export("provider_configured", True)

# Note: For managing members, this Members resource provides a declarative
# way to set the *entire* list of members per role.
# Use with caution, as it will overwrite existing members not specified here.

# Similarly, the Group resource manages the member list declaratively. 